rfc/plugins.md

```diff
@@ -0,0 +1,300 @@
+
+However, for any remotely-capable extension, its contract must be defined as protobufs, rather than exposed as simple POJOs or interfaces. A key distinction is that we’re not specifying gRPC as the transport.  A remote transport may be Amazon SQS or Redis PubSub, not necessarily gRPC.
+
+### Technology Compatibility Kits
+_alias: TCKs_
+
+A convention of writing TCKs for each Extension Point will be established. The concept of a TCK is borrowed from the JSR world, where compatibility tests are written which assert correct behavior of various features.   Each implementation must them pass those compatibility tests.  For Spinnaker plugins, TCKs can be written for two different consumers: Extension Point providers and Plugin developers.
+
+TCKs assert correct behavior of an extension point and its associated plugin(s). For any extension point, a TCK will be written that asserts even pathological plugins behave in an expected way while integrating with the extension point. For a plugin, a TCK will be provided that a plugin developer can use to assert correctness without necessarily needing to run other Spinnaker services.
+
+This goes further than the convention of using Protobuf to define the interfaces of a particular plugin, where Protobufs only define the interface, whereas a TCK will assert behavior that lies beneath the interface. Combining a TCK and adherence to the interface, contracts will be fully asserted between a plugin and its extension point(s).
+
+### Up/Downgrading Plugins
+This section applies to federated development scenarios, where the operator is not necessarily the same as the plugin developers; where the release cadence of a plugin does not adhere to the release cadence of the core Spinnaker services.
+
+The upgrading (or downgrading, from now on broadly upgrading) plugin story is different based on the particular plugin runtime.
+
+#### In-process
+When a service starts up, PF4J will scan a designated locally-accessible directory for plugins to install (these plugins may be downloaded at runtime or baked into the image at built-time). Once an application has started, it will notify Front50 of the plugins that are installed, and get a response dictating which plugins are to be enabled.
+
+In the case of a runtime-downloaded plugin, version to download will be dictated by Front50. Should Front50 be unavailable, the service will download the version defined by the service, or if none is defined, the latest plugin version that matches the Spinnaker service version.
+
+When a new plugin version is released, it is the responsibility of the operator to release this new version. _**TBD**: How will an operator know a new version exists?_ This new version will need to be released in lock-step with the services that it is associated with. In the future, this functionality _may_ be dynamically sourced and then re-activated. See [pf4j/pf4j-update](https://github.com/pf4j/pf4j-update)
```

Comment:
> _**TBD**: How will an operator know a new version exists?

The Jenkins model is to have an official plugin repository from which metadata about available plugins can be queried and displayed on the UI. Not that we want to follow Jenkins 😃, but it occurred to me that mentioning a plugin repository as a goal/non-goal might be relevant as well.